### PR TITLE
[libSyntax] Support parsing several type kinds

### PR DESCRIPTION
**Pull request by rintaro**
* ArrayType (`[Elem]`)
* DictionaryType (`[Key: Val]`)
* OptionalType (`?`)
* ImplicitlyUnwrappedOptionalType (`!`)
* MetatypeType (`.Type`, `.Protocol`)

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/Parse/ParseType.cpp [hunk 1]
```diff
@@ -249,24 +253,46 @@ ParserResult<TypeRepr> Parser::parseTypeSimple(Diag<> MessageID,
...
         ty = makeParserResult(ty,
           new (Context) ProtocolTypeRepr(ty.get(), protocolLoc));
+        if (SyntaxContext->isEnabled()) {
+          MetatypeTypeSyntaxBuilder Builder;
+          Builder
+            .useTypeOrProtocol(SyntaxContext->popToken())
+            .usePeriod(SyntaxContext->popToken())
+            .useBaseType(SyntaxContext->popIf<TypeSyntax>().getValue());
+          SyntaxContext->addSyntax(Builder.build());
+        }
```
Comment (nkcsgexi) [inline lib/Parse/ParseType.cpp]: Can we extract a function for the above block to reduce dupe?

**Review by nkcsgexi — COMMENTED** [review]
Looking good! One minor comment.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Pull request merged**